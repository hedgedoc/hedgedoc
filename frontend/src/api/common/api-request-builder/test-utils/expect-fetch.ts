/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { defaultConfig } from '../../default-config'
import { Mock } from 'ts-mockery'

/**
 * Mock fetch api for tests.
 * Check that the given url and options are present in the request and return the given status code.
 *
 * @param expectedUrl the url that should be requested
 * @param requestStatusCode the status code the mocked request should return
 * @param expectedOptions additional options
 */
export const expectFetch = (
  expectedUrl: string,
  requestStatusCode: number,
  expectedOptions: RequestInit,
  responseBody?: unknown
): void => {
  global.fetch = jest.fn((fetchUrl: RequestInfo | URL, fetchOptions?: RequestInit): Promise<Response> => {
    expect(fetchUrl).toEqual(expectedUrl)
    expect(fetchOptions).toStrictEqual({
      ...defaultConfig,
      body: undefined,
      headers: new Headers(),
      ...expectedOptions
    })
    return Promise.resolve(
      Mock.of<Response>({
        status: requestStatusCode,
        statusText: mapCodeToText(requestStatusCode),
        json: jest.fn(() => (responseBody ? Promise.resolve(responseBody) : Promise.reject()))
      })
    )
  }) as typeof global.fetch
}
const mapCodeToText = (code: number): string => {
  switch (code) {
    case 400:
      return 'bad_request'
    case 401:
      return 'forbidden'
    default:
      return 'unknown_code'
  }
}
