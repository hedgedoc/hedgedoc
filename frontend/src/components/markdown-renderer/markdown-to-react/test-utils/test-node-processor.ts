/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { NodeProcessor } from '../../node-preprocessors/node-processor'
import type { Document } from 'domhandler'
import { Element, Text } from 'domhandler'

export class TestNodeProcessor extends NodeProcessor {
  process(document: Document): Document {
    document.childNodes.push(new Element('node-processor', {}, [new Text('node processor children')]))
    return document
  }
}
