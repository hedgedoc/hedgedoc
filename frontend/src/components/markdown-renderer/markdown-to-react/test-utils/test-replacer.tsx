/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { NativeRenderer, NodeReplacement, SubNodeTransform } from '../../replace-components/component-replacer'
import { ComponentReplacer, DO_NOT_REPLACE } from '../../replace-components/component-replacer'
import type { Element } from 'domhandler'
import React, { Fragment } from 'react'

export class TestReplacer extends ComponentReplacer {
  replace(node: Element, subNodeTransform: SubNodeTransform, nativeRenderer: NativeRenderer): NodeReplacement {
    return node.tagName === 'node-processor' ? (
      <Fragment>
        <span>NodeProcessor! </span>
        <span data-children={true}> {node.childNodes.map(subNodeTransform)} </span>
        <span data-native={true}> {nativeRenderer()} </span>
      </Fragment>
    ) : (
      DO_NOT_REPLACE
    )
  }
}
