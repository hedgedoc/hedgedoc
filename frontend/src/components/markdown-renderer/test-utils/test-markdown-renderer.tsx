/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { StoreProvider } from '../../../redux/store-provider'
import type { MarkdownRendererExtension } from '../extensions/_base-classes/markdown-renderer-extension'
import { MarkdownToReact } from '../markdown-to-react/markdown-to-react'
import React, { useMemo } from 'react'

export interface SimpleMarkdownRendererProps {
  content: string
  extensions: MarkdownRendererExtension[]
}

/**
 * A markdown renderer for tests.
 *
 * @param content The content to be rendered.
 * @param extensions The {@link MarkdownRendererExtension MarkdownExtensions} to use for rendering.
 */
export const TestMarkdownRenderer: React.FC<SimpleMarkdownRendererProps> = ({ content, extensions }) => {
  const lines = useMemo(() => content.split('\n'), [content])
  return (
    <StoreProvider>
      <MarkdownToReact
        markdownContentLines={lines}
        markdownRenderExtensions={extensions}
        newlinesAreBreaks={true}
        allowHtml={false}
      />
    </StoreProvider>
  )
}
