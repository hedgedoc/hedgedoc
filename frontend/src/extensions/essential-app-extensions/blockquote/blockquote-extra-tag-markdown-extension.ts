/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/_base-classes/markdown-renderer-extension'
import type { NodeProcessor } from '../../../components/markdown-renderer/node-preprocessors/node-processor'
import type { ComponentReplacer } from '../../../components/markdown-renderer/replace-components/component-replacer'
import { BlockquoteBorderColorNodePreprocessor } from './blockquote-border-color-node-preprocessor'
import { BlockquoteColorExtraTagReplacer } from './blockquote-color-extra-tag-replacer'
import { BlockquoteExtraTagMarkdownItPlugin } from './blockquote-extra-tag-markdown-it-plugin'
import { BlockquoteExtraTagReplacer } from './blockquote-extra-tag-replacer'
import type MarkdownIt from 'markdown-it'

/**
 * Adds support for generic blockquote extra tags and blockquote color extra tags.
 */
export class BlockquoteExtraTagMarkdownExtension extends MarkdownRendererExtension {
  public static readonly tagName = 'app-blockquote-tag'

  public configureMarkdownIt(markdownIt: MarkdownIt): void {
    new BlockquoteExtraTagMarkdownItPlugin('color', 'tag').registerRule(markdownIt)
    new BlockquoteExtraTagMarkdownItPlugin('name', 'person').registerRule(markdownIt)
    new BlockquoteExtraTagMarkdownItPlugin('time', 'clock').registerRule(markdownIt)
  }

  public buildReplacers(): ComponentReplacer[] {
    return [new BlockquoteColorExtraTagReplacer(), new BlockquoteExtraTagReplacer()]
  }

  public buildNodeProcessors(): NodeProcessor[] {
    return [new BlockquoteBorderColorNodePreprocessor()]
  }

  public buildTagNameAllowList(): string[] {
    return [BlockquoteExtraTagMarkdownExtension.tagName]
  }
}
