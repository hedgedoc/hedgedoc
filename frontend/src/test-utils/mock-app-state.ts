/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import * as useApplicationStateModule from '../hooks/common/use-application-state'
import type { ApplicationState } from '../redux'
import { initialState as initialStateDarkMode } from '../redux/dark-mode/initial-state'
import { initialState as initialStateEditorConfig } from '../redux/editor-config/initial-state'
import { initialState as initialStateNoteDetails } from '../redux/note-details/initial-state'
import { initialState as initialStateRealtimeStatus } from '../redux/realtime/initial-state'
import { initialState as initialStateRendererStatus } from '../redux/renderer-status/initial-state'
import type { NoteDetails } from '../redux/note-details/types'
import type { RealtimeStatus } from '../redux/realtime/types'
import type { DeepPartial } from '@hedgedoc/commons'

jest.mock('../redux/editor-config/methods', () => ({
  loadFromLocalStorage: jest.fn().mockReturnValue(undefined)
}))
jest.mock('../hooks/common/use-application-state')

/**
 * Mocks the {@link ApplicationState} for a test.
 * When not overriden, it uses the initial state of the reducers.
 *
 * @param state Overrides for the mocked state
 */
export const mockAppState = (state?: DeepPartial<ApplicationState>) => {
  jest.spyOn(useApplicationStateModule, 'useApplicationState').mockImplementation((fn) => {
    return fn({
      darkMode: {
        ...initialStateDarkMode,
        ...state?.darkMode
      },
      editorConfig: {
        ...initialStateEditorConfig,
        ...state?.editorConfig
      },
      history: state?.history ?? [],
      noteDetails: {
        ...initialStateNoteDetails,
        ...state?.noteDetails
      } as NoteDetails,
      realtimeStatus: {
        ...initialStateRealtimeStatus,
        ...state?.realtimeStatus
      } as RealtimeStatus,
      rendererStatus: {
        ...initialStateRendererStatus,
        ...state?.rendererStatus
      },
      user: state?.user ?? null
    })
  })
}
