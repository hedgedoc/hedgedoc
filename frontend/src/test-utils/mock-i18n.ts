/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { TFunction } from 'i18next'
import i18n from 'i18next'
import { initReactI18next } from 'react-i18next'

/**
 * Initializes i18n with minimal settings and without any data, so it just returns the used key as translation.
 *
 * @return A promise that resolves if i18n has been initialized
 */
export const mockI18n = (): Promise<TFunction> => {
  return i18n.use(initReactI18next).init({
    lng: 'en',
    fallbackLng: 'en',
    ns: ['translationsNS'],
    defaultNS: 'translationsNS',
    interpolation: {
      escapeValue: false
    },
    resources: { en: { translationsNS: {} } }
  })
}
