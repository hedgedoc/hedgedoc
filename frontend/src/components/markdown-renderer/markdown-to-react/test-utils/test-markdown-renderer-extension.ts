/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { MarkdownRendererExtension } from '../../extensions/_base-classes/markdown-renderer-extension'
import type { NodeProcessor } from '../../node-preprocessors/node-processor'
import type { ComponentReplacer } from '../../replace-components/component-replacer'
import { TestNodeProcessor } from './test-node-processor'
import { TestReplacer } from './test-replacer'
import type MarkdownIt from 'markdown-it'
import Token from 'markdown-it/lib/token'

export class TestMarkdownRendererExtension extends MarkdownRendererExtension {
  buildNodeProcessors(): NodeProcessor[] {
    return [new TestNodeProcessor()]
  }

  configureMarkdownIt(markdownIt: MarkdownIt) {
    markdownIt.use(() => {
      markdownIt.core.ruler.push('configure', (core) => core.tokens.push(new Token('configure', 'configure', 0)))
      markdownIt.renderer.rules.configure = () => '<span>configure</span>'
    })
  }

  buildReplacers(): ComponentReplacer[] {
    return [new TestReplacer()]
  }

  buildTagNameAllowList(): string[] {
    return ['node-processor']
  }

  configureMarkdownItPost(markdownIt: MarkdownIt) {
    markdownIt.use(() => {
      markdownIt.core.ruler.push('post', (core) => core.tokens.push(new Token('post', 'post', 0)))
      markdownIt.renderer.rules.post = () => '<span>post</span>'
    })
  }
}
