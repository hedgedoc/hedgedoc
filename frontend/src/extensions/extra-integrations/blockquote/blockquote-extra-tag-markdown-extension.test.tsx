/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { TestMarkdownRenderer } from '../../../components/markdown-renderer/test-utils/test-markdown-renderer'
import { BlockquoteExtraTagMarkdownExtension } from './blockquote-extra-tag-markdown-extension'
import { render } from '@testing-library/react'
import React from 'react'

describe('blockquote extra tag', () => {
  ;[
    '[color=white]',
    '[color=#dfe]',
    '[color=notarealcolor]',
    '[color=#abcdef]',
    '[color=]',
    '[name=giowehg]',
    '[name=]',
    '[time=tomorrow]',
    '[time=]',
    '[key]',
    '[key=]',
    '[=value]',
    '> [color=#f00] text'
  ].forEach((content) => {
    it(`renders the tag "${content}" correctly`, () => {
      const view = render(
        <TestMarkdownRenderer extensions={[new BlockquoteExtraTagMarkdownExtension()]} content={content} />
      )
      expect(view.container).toMatchSnapshot()
    })
  })
})
