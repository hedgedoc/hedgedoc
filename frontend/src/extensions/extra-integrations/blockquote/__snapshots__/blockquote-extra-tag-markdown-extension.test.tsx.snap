// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blockquote extra tag renders the tag "> [color=#f00] text" correctly 1`] = `
<div>
  <blockquote
    style="border-left-color: #f00;"
  >
    

    <p>
       text
    </p>
    

  </blockquote>
  

</div>
`;

exports[`blockquote extra tag renders the tag "[=value]" correctly 1`] = `
<div>
  <p>
    [=value]
  </p>
  

</div>
`;

exports[`blockquote extra tag renders the tag "[color=#abcdef]" correctly 1`] = `
<div>
  <p>
    <span
      class="blockquote-extra"
      style="color: rgb(171, 205, 239);"
    >
      <i
        class="fa  fa-tag mx-1 "
      />
    </span>
  </p>
  

</div>
`;

exports[`blockquote extra tag renders the tag "[color=#dfe]" correctly 1`] = `
<div>
  <p>
    <span
      class="blockquote-extra"
      style="color: rgb(221, 255, 238);"
    >
      <i
        class="fa  fa-tag mx-1 "
      />
    </span>
  </p>
  

</div>
`;

exports[`blockquote extra tag renders the tag "[color=]" correctly 1`] = `
<div>
  <p>
    [color=]
  </p>
  

</div>
`;

exports[`blockquote extra tag renders the tag "[color=notarealcolor]" correctly 1`] = `
<div>
  <p>
    <span
      class="blockquote-extra"
    >
      <i
        class="fa  fa-tag mx-1 "
      />
      notarealcolor
    </span>
  </p>
  

</div>
`;

exports[`blockquote extra tag renders the tag "[color=white]" correctly 1`] = `
<div>
  <p>
    <span
      class="blockquote-extra"
      style="color: white;"
    >
      <i
        class="fa  fa-tag mx-1 "
      />
    </span>
  </p>
  

</div>
`;

exports[`blockquote extra tag renders the tag "[key=]" correctly 1`] = `
<div>
  <p>
    [key=]
  </p>
  

</div>
`;

exports[`blockquote extra tag renders the tag "[key]" correctly 1`] = `
<div>
  <p>
    [key]
  </p>
  

</div>
`;

exports[`blockquote extra tag renders the tag "[name=]" correctly 1`] = `
<div>
  <p>
    [name=]
  </p>
  

</div>
`;

exports[`blockquote extra tag renders the tag "[name=giowehg]" correctly 1`] = `
<div>
  <p>
    <span
      class="blockquote-extra"
    >
      <i
        class="fa  fa-user mx-1 "
      />
      giowehg
    </span>
  </p>
  

</div>
`;

exports[`blockquote extra tag renders the tag "[time=]" correctly 1`] = `
<div>
  <p>
    [time=]
  </p>
  

</div>
`;

exports[`blockquote extra tag renders the tag "[time=tomorrow]" correctly 1`] = `
<div>
  <p>
    <span
      class="blockquote-extra"
    >
      <i
        class="fa  fa-clock-o mx-1 "
      />
      tomorrow
    </span>
  </p>
  

</div>
`;
